Add [default] wallet encryption

halseth opened the pull request:
This PR makes it possible to encrypt `lnd`'s wallet database ~~by specifying `--encryptwallet`~~ (on by default, disable by `--noencryptwallet`) at startup. This will set up a simple gRPC service that can handle setting a password at initial wallet creation, or receive a password for decrypting an existing wallet. When this is received, `lnd` will continue running as normal. The password service is only operational at startup.

#### Startup
`lnd` with non-existing `wallet.db`

#### Usage
`lncli create` will ask the user to input a password + confirm it, then send the password over RPC to `lnd` for initial wallet creation.
`lncli unlock` will ask for password and send it to `lnd` over RPC for unlocking the wallet.
Both `lncli` commands has strong terminal replay protection through their use of the secure password read method.

### Things to consider
The current implementation was kept as simple as possible, to check the viability of this approach. Things to consider adding in the future (or in this PR):
- ~~Retries in case of wrong password~~ **Added**
- A stop/cancel command. `lnd` will currently wait for the password to be provided, and since the normal rpcServer is not active, `lncli stop` will not work.
- A return value (success/error) that can be printed as output from `lncli password`
- ~~REST support~~ **Added**
- Change password functionality. As of now the password cannot be changed after chosen at first startup.
- ~~Password confirmation. At first startup, we should consider requiring the user to input the password twice, to confirm the input.~~ **Added**
- Interactive setup (seed output etc). We can guide the user through a setup process, making them write down seed, input password etc.
- ~~Macaroons?~~ Added, but this needs some more love when we also encrypt the macaroon db.

Will need to update `btcwallet` to newest version to avoid database freeze after entering one password.

Closes #226.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -25,6 +25,21 @@ package lnrpc;
  */
 
+// The Password service is used to provide a wallet database password for lnd.
+service Password {
```
Naming suggestion: `DaemonUnlocker`?

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -25,6 +25,21 @@ package lnrpc;
+    rpc EncryptionPassword(EncryptionPasswordRequest) returns (EncryptionPasswordResponse) {}
+}
+
+message EncryptionPasswordRequest {
```
Naming suggestion: `UnlockDaemonRequest`.

Roasbeef commented:
> REST support

We gain REST support easily by just adding the necessary annotations to the proto file itself, just like all the other calls on the current sole gRPc service. 

> Change password functionality. As of now the password cannot be changed after chosen at first startup.

This isn't _yet_ compatible with the way the default `WalletController` implementation (`btcwallet`) handles wallet encryption. It wouldn't be tooo difficult to implement, but we'd need to ensure that the switch over was done atomically, as it involves _re-encrypting_ all database state. 

> Interactive setup (seed output etc). We can guide the user through a setup process, making them write down seed, input password etc.

With this in mind, I think we should split the service into two methods: `Create` and `Unlock`. `Create` can optionally also take an HD seed that the user provides themselves (w/o any full restore+scan+lookahead logic for now). This will be useful, as GUI's can then implement BIP 39 and use that to generate the HD seed that's passed in. 

> Macaroons?

The passphrase can be used along with a PKDF to encrypt the macaroon database which itself is currently plaintext (and shouldn't be as it stores the root keys).

halseth commented on passwordservice/service.go:
```diff
@@ -0,0 +1,31 @@
+func (p *PasswordService) EncryptionPassword(ctx context.Context,
+	in *lnrpc.EncryptionPasswordRequest) (*lnrpc.EncryptionPasswordResponse,
+	error) {
+	p.Password <- in.Password
```
This is absolutely something I wanted to do, but was unsure how to determine if the password is correct before trying to instantiate the `ChainControl`. 

I will look more into this, as that will make the service much more flexible in regards to retries/confirmations etc, and also for testability as you mention.

halseth replied:
This is now done, the package tries opening the database before returning the password to the daemon, to determine if it's correct.

halseth commented:
Thanks for the comments! Will restructure and rename the service, and look into whether the password checking can be made fully contained within the package.

Regarding REST support, I will add that, trying to extract the common setup code between the two services. I was worried the security of passing the password over REST would be lower, but it's okay since we enforce TLS and we can do POST requests?

Will also create the `create/unlock` methods, adding the suggested macaroon authentication.

halseth commented on config.go:
```diff
@@ -126,6 +127,8 @@ type config struct {
 
 	NoNetBootstrap bool `long:"nobootstrap" description:"If true, then automatic network bootstrapping will not be attempted."`
+
+	EncryptWallet bool `long:"encryptwallet" description:"Require client to provide password for wallet encryption."`
```
Will make this default or enforce it when the general design is agreed upon. I suspect I will have to change some integration tests...

jimpo commented on passwordservice/service.go:
```diff
@@ -0,0 +1,133 @@
+func (p *PasswordService) CreateWallet(ctx context.Context,
+	in *lnrpc.CreateWalletRequest) (*lnrpc.CreateWalletResponse, error) {
+
+	cfg := &btcwallet.Config{
```
What is the purpose of creating this `cfg`?

halseth replied:
removed

jimpo commented on lnd.go:
```diff
@@ -119,10 +120,113 @@ func lndMain() error {
+	// is not set, we use the default password "hello" for wallet
+	// encryption.
+	walletPw := []byte("hello")
+	if cfg.EncryptWallet {
```
The contents of this `if` looks like it can cleanly be broken into a separate method `getWalletPasswordFromUser` or something.

halseth replied:
good idea! done :)

jimpo commented on lnd.go:
```diff
@@ -119,10 +120,113 @@ func lndMain() error {
+		 * primitives, but the client's preference order has more
+		 * effect during negotiation.
+		**/
+		CipherSuites: []uint16{
```
Maybe define the ciphers list as a constant at the top of the file?

jimpo commented on lnd.go:
```diff
@@ -119,10 +120,113 @@ func lndMain() error {
+		// used for getting the encryption password from the client.
+		lis, err := net.Listen("tcp", grpcEndpoint)
+		if err != nil {
+			fmt.Printf("failed to listen: %v", err)
```
Use `rpcsLog` or just return this as `return fmt.Errorf(...)`

halseth replied:
`fmt.Printf` is used similarly elsewhere in this file, so maybe there's a reason for that? cc @Roasbeef

halseth commented:
Made wallet encrypted by default, changed name of service to `UnlockerService`, and added unit tests.

REST support is also added, and now the two rpc calls `lncli create` and `lncli unlock` is now used to create and unlock the wallet.

jimpo commented on walletunlocker/service_test.go:
```diff
@@ -0,0 +1,142 @@
+	t.Parallel()
+
+	// testDir is empty, meaning wallet was not created from before.
+	testDir, err := ioutil.TempDir("", "testcreate")
```
Clean up temp dir with a defer?

jimpo commented on walletunlocker/service.go:
```diff
@@ -0,0 +1,127 @@
+		netParams:       params,
+	}
+}
+func (u *UnlockerService) CreateWallet(ctx context.Context,
```
nit: Add newline above

halseth commented:
Fixed the last nits, cleaned up commit history and rebased on master. PTAL :)

halseth commented on walletunlocker/service.go:
```diff
@@ -0,0 +1,132 @@
+	in *lnrpc.CreateWalletRequest) (*lnrpc.CreateWalletResponse, error) {
+
+	// Check macaroon to see if this is allowed.
+	if u.authSvc != nil {
```
I moved the macaroon generation logic above the code initiating the `UnlockerService`, so think they should be present at that point.
